### PR TITLE
Use metadata in CSI

### PR DESCRIPTION
**Pull request by jichenjc**
use util/metadata in CSI 

**What this PR does / why we need it**:
CSI need use util/metadata directly, instead of using its own

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
The metadata utils was created and all openstack metadata related routine use same functions now.
```

### CODE REVIEW

**Comment by adisky** [comment]
/hold 
@jichenjc unit test needs to be fixed
https://github.com/kubernetes/cloud-provider-openstack/pull/487#issuecomment-464625101

**Comment by dims** [comment]
hmm, looks like @adisky comment has not been addressed - https://github.com/kubernetes/cloud-provider-openstack/pull/487#issuecomment-464983795

**Comment by jichenjc** [comment]
need some update for the metadata service
/wip

File: pkg/util/metadata/metadata.go [hunk 1]
```diff
@@ -205,3 +209,21 @@ func Get(order string) (*Metadata, error) {
 	}
 	return metadataCache, nil
 }
+
+// GetInstanceID from metadata service
```
Comment (ramineni) [inline pkg/util/metadata/metadata.go]: needs to be updated, as now it retrieves from both metadata servie and config drive

File: pkg/util/metadata/metadata.go [hunk 2]
```diff
@@ -205,3 +209,21 @@ func Get(order string) (*Metadata, error) {
...
+func GetInstanceID(order string) (string, error) {
+	md, err := Get(order)
+	if err != nil {
+		return "", err
+	}
+	return md.UUID, nil
+}
+
+// GetAvailabilityZone returns zone from metadata service
```
Comment (ramineni) [inline pkg/util/metadata/metadata.go]: ditto

File: pkg/util/metadata/metadata.go (hunk 1 above)
Comment (jichenjc) [reply pkg/util/metadata/metadata.go]: ok

File: pkg/util/metadata/metadata.go [hunk 3]
```diff
@@ -101,6 +101,10 @@ func getConfigDrivePath(metadataVersion string) string {
 	return fmt.Sprintf(configDrivePathTemplate, metadataVersion)
 }
```
Comment (adisky) [inline pkg/util/metadata/metadata.go]: @jichenjc  I think we should have an interface for metadata package, it is required for testing, and while doing csi sanity testing for cinder csi driver i realized  it is must

Comment (jichenjc) [reply pkg/util/metadata/metadata.go]: @adisky I have similar thought, let me try to work on it soon

File: pkg/util/metadata/metadata.go [hunk 4]
```diff
@@ -205,3 +211,31 @@ func Get(order string) (*Metadata, error) {
...
+	md, err := m.Get(order)
+	if err != nil {
+		return "", err
+	}
+	return md.AvailabilityZone, nil
+}
+
+var IMetadataInterface IMetadata = nil
+
+func GetMetadata() IMetadata {
```
Comment (adisky) [inline pkg/util/metadata/metadata.go]: GetMetadata is confusing, GetMetadataProvider (as used before) or GetMetadataInstance is better

File: pkg/util/metadata/metadata.go [hunk 5]
```diff
@@ -77,6 +77,15 @@ type Metadata struct {
 	// .. and other fields we don't care about.  Expand as necessary.
 }
 
+type IMetadata interface {
+	GetInstanceID(order string) (string, error)
```
Comment (adisky) [inline pkg/util/metadata/metadata.go]: GetInstanceID, GetAvailabilityZone this is not necessary, we can modify csi code to get metadata as ccm is doing, to have consistency.

File: pkg/util/metadata/metadata.go [hunk 6]
```diff
@@ -77,6 +77,15 @@ type Metadata struct {
...
+	GetAvailabilityZone(order string) (string, error)
+	Get(order string) (*Metadata, error)
```
Comment (adisky) [inline pkg/util/metadata/metadata.go]: here, GetMetadata is better then Get

File: pkg/util/metadata/metadata.go [hunk 7]
```diff
@@ -77,6 +77,15 @@ type Metadata struct {
...
+	GetFromMetadataService(metadataVersion string) (*Metadata, error)
+	GetFromConfigDrive(metadataVersion string) (*Metadata, error)
+	GetDefaultMetadataSearchOrder() string
```
Comment (adisky) [inline pkg/util/metadata/metadata.go]: IMO GetDefaultMetadataSearchOrder(), it can be a utility method only need not to be part of interface

Comment (jichenjc) [reply pkg/util/metadata/metadata.go]: I need double check this and see whether it's ok to do so, some place seems rely on this function

**Comment by huangtianhua** [comment]
recheck

**Comment by huangtianhua** [comment]
recheck

**Comment by jichenjc** [comment]
I was informed this stuck the test ,so close this and will reopen soon

**Comment by adisky** [comment]
/close
please reopen when you start working on it again

**Pull request closed without merging**